File: java/src/ServerThread.java
import java.net.*;
import java.io.*;
class ServerThread extends Thread {
    private Socket client;
    private PrintWriter out;
    private JobQueue queue;

    ServerThread(Socket client, JobQueue queue){
        super("HandBrake Network Queue Server Thread");
        this.client = client;
        this.queue = queue;
    }

    public void run(){
        try {
            // Java stupidity declaring objects for objects
            out = new PrintWriter( client.getOutputStream(), true);
        } catch(IOException e){
            System.err.println("Failed to allocate resources to use Socket");
            System.err.println(e);
            System.exit(-2);
        }

        String inputLine;
        while ((inputLine = FileTransfer.readLine(client)) != null){
            switch(Command.getCommandType(inputLine)){
            case GET_JOB:
                getJob();
                break;
            case CHECKOUT_JOB:
                checkoutJob();
                break;
            case COMPLETE_JOB:
                completeJob();
                break;
            default:
                System.err.println("Error: Unkown Command: " + inputLine);
                System.err.println("Continuing...");
            }
        }
        clean();
    }

    private void clean() {
        try {
            out.close();
            client.close();
        } catch (IOException e) {
            System.err.println("Error when closing resources from thread");
        }
    }

    private void getJob() {
        System.out.println( client.getRemoteSocketAddress() + " has requested a job");		
        Job job = queue.getNextJob();
        if(job == null){
            // Tell them that we have no job
            out.println(Command.NO_JOB);
        }else{
            out.println(Command.HAVE_JOB);
        }
        try {
            Job.send(new PrintWriter(client.getOutputStream()), job);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void checkoutJob() {
        // The first thing we read is the ID of the job being checked out
        int id = readID();
        System.out.println(client.getRemoteSocketAddress() + " has checked out job ID " + id);
        
        Job job = queue.checkout(id);
        FileTransfer.sendDirs(client, job.relativeDirsWithSourceDir());
        
        FileTransfer.send(client, job.relativeFilesWithSourceDir());
    }

    private void completeJob() {
        // The first thing we read is the ID of the job being checked out
        int id = readID();
        FileTransfer.recv(client);
        queue.complete(id);
        System.out.println(client.getRemoteSocketAddress() + " has completed job ID " + id);
    }

    private int readID(){
        try {
            return Integer.parseInt(FileTransfer.readLine(client));
        } catch (NumberFormatException e) {
            System.err.println("Unable to parse ID when reading from socket");
            System.err.println("Continuing...");
        }
        return -1;
    }
}

File: ruby/lib/server_thread.rb
# vi: et sw=4
require "./lib/file_transfer"
class ServerThread
    def initialize(client, queue)
        @client = client
        @queue = queue
    end

    def serve_client
        command = @client.gets.chomp
        puts command
        case command 
        when Command::GET_JOB
            get_job()
        when Command::CHECKOUT_JOB
            checkout_job()
        when Command::COMPLETE_JOB
            complete_job()
        when ""
            # They closed the socket
        else
            STDERR.puts "unknown command: #{command.inspect}"
            @client.puts(Command::UNKNOWN)
        end
    ensure
        @client.close
    end
    alias :run :serve_client

    private
    def get_job
        puts "client has requested a job"
        job = @queue.next_job()
        if job
            @client.puts(Command::HAVE_JOB)
            Job.send(@client,job)
            puts "Sent job ID #{job.id}"
        else
            @client.puts(Command::NO_JOB)
            STDERR.puts "No new jobs left..."
        end
    end

    def checkout_job
        id = read_id()
        job = @queue.checkout(id)
        puts "client has checked out job id #{id}"

        # Send them the directory structure
        FileTransfer.send_dirs(@client, job.relative_dirs_with_source_dir)

        # Save where we are
        pwd = Dir.pwd
        # Go to the source folder so we can send relative paths
        Dir.chdir(job.source + "/..")

        # Get the files required
        FileTransfer.send(@client,job.relative_files_with_source_dir)

        # Go back to where we were
        Dir.chdir(pwd)
    end

    def complete_job
        id = read_id()
        puts "client has completed a job id: #{id}"
        FileTransfer.recv(@client)
        @queue.complete(id)
    end

    def read_id
        @client.gets
    end
end
